draft-ietf-tls-external-psk-importer.md

```diff
@@ -138,14 +178,46 @@ EPSKs may be imported for early data use if they are bound to protocol settings
 otherwise be required for early data with normal (ticket-based PSK) resumption. Minimally, that means ALPN,
 QUIC transport settings, etc., must be provisioned alongside these EPSKs.
 
+##Client ID {#client-id-description}
+This is to prevent Selfie-style reflections.
+This is only necessary in scenarios where more than two actors use the same key, including the case where a single agent will complete PSK handshakes as both the client and the server using the same key.
+The Selfie attack {{!Selfie}} abuses an underlying assumption of TLS, that only one client and one server know a PSK.
+A pair of agents that both act as both client *and* server, and use the same PSK in both roles violate this assumption.
+In this case, there are two clients and two servers that know the PSK, since each endpoint acts as both client and server.
+The agents are thus both vulnerable to reflection attacks where the attacker forces the agent to act as both client and server in a single connection.
+This is because neither agent can distinguish between itself and its peer.
+Recall both servers have the exact same configuration.
+Therefore even though the session specific parameters change for each session, the client cannot distinguish between a response reflected back on it by the attacker and an honest response from its peer.
+For each parameter, both servers either return the value for every session, or they return a different value for every session.
+By hashing a distinguishing `client_id` value into the PSK binder, a server that received a reflected `ClientHello` would be unable to verify the binder, and would reject the connection.
+
+For example, distinguishing strings for agents in peer-to-peer IoT device deployments could be a MAC address.
+Similarly, distinguishing strings for agents in mesh VM deployments could be namespaces.
+The decision to use this field and what it should contain MUST be agreed OOB.
+Note that, critically, this value is never sent on the wire, as this would identify the client even to passive adverseries.
```

Comment:
Some issues:
- This whole text seems rather bloated, can it be written in a more concise way?
- The last sentence is not necessarily true. If a MAC address is used as client identifier, then it will be visible somewhere on the wire.
- The (non-)requirements for "client_id" is not explicitly specified. Does it have to be globally unique? What if this "requirement" is not satisfied?

The Selfie issue occurred because a PSK was reused for a client and server role. What about a simpler solution: require implementations to store the role with the PSK identity? If desired, this role (1 byte) could be explicitly included in the context.

Reply:
W.r.t. the Selfie issue, including the role was my initial thought, however if you think it through it actually doesn't help at all. Agent 1, acting in its role as Client sends a `ClientHello` with the role "client". The attacker reflects the `ClientHello` back at Agent 1. Agent 1, receives the `ClientHello`, and because it expects Agent 2 to use the role "client" when it acts as the client, accepts. 
The problem is specifically that both agents act completely identically, so they are indistinguishable. 
In fact, in this configuration the agents don't even know which agent they are. 

1. The text is trying to refute exactly the argument you made. It probably needs to be longer. 

2. I agree that the sentence is overbroad, I'll have to be more precise.

3. Good catch. The requirements are listed in a previous section, but the link got lost as I shuffled the text around.